### PR TITLE
[Fix #3456]Support multiple constraint over same connection

### LINKED ISSUE #3456: Serverless workflow switch issue: when two branches pointing to the same transition, it treats the conditions as "and"
### Describe the bug

I'm using serverless workflow, in a "switch" state, I defined two branches pointing to the same transitions. I'm expecting the "switch" will route the workflow to the first condition matched. However, the actual match is "switch" seems treat two branches as "and", which unless both conditions get matched, switch won't jump to the expected branch.

### Expected behavior

Output

```
{workflowdata={"condition1":true,"condition2":false,"output":"branch1"}}
```

### Actual behavior

Output

```
{workflowdata={"condition1":true,"condition2":false,"output":"default"}}
```

### How to Reproduce?

Workflow Yaml file

```yaml
id: switch
version: 1.0.0
specVersion: "0.8"
name: Switch test
start: switch test
states:
  - name: switch test
    type: switch
    dataConditions:
      - name: Branch1-Name
        condition: .condition1
        transition: branch1
      - name: Branch2-Name
        condition: .condition2
        transition: branch1

    defaultCondition:
      transition: default

  - name: branch1
    type: inject
    data:
      output: branch1
    end: true


  - name: default
    type: inject
    data:
      output: default
    end: true
 
```

Test java code

```java
@Test
public void testKogitoSwitch() throws Exception {
    final Map<String, Object> input = ImmutableMap.of(
            "condition1", true,
            "condition2", false);

    try (StaticWorkflowApplication application = StaticWorkflowApplication.create()) {
        var process = application.process(getWorkflowFromResource("workflow/switch.sw.yaml"));
        var processInstance = process.createInstance(new JsonNodeModel(input));
        processInstance.start();
        final JsonNodeModel model = processInstance.variables();
        System.out.println(model.toMap());
    }
}
```

If change the input as

```java
   final Map<String, Object> input = ImmutableMap.of(
            "condition1", true,
            "condition2", true);
```

Output would be
```
{workflowdata={"condition1":true,"condition2":true,"output":"branch1"}}
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Kogito version or git rev (or at least Quarkus version if you are using Kogito via Quarkus platform BOM)

2.44.0.Alpha, 1.44.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by fjtirado**
Fixes https://github.com/apache/incubator-kie-kogito-runtimes/issues/3456
The engine was not ready to support multiple constraints over the same connection. That makes sense for BPMN, but it is not necesarily true for SWF, as shown by the example in the issue.
Therefore, it has been changed to support both  (BPMN constraint will be size one)

### CODE REVIEW

**Comment by kie-ci3** [comment]
**PR job** `#1` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3459 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/1/display/redirect

**Test results:**
- PASSED: 1331
- FAILED: 2

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/1/testReport/org.jbpm.bpmn2/FlowTest/testExclusiveSplitWithNoConditions/">org.jbpm.bpmn2.FlowTest.testExclusiveSplitWithNoConditions</a></summary>
Should fail as XOR gateway does not have conditions defined
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/1/testReport/org.jbpm.bpmn2.canonical/ActivityGenerationModelTest/testInclusiveSplitDefaultConnection/">org.jbpm.bpmn2.canonical.ActivityGenerationModelTest.testInclusiveSplitDefaultConnection</a></summary>
<br/>expected: 2<br/> but was: 5
</details>

**Comment by kie-ci3** [comment]
**PR job** `#5` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3459 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/5/display/redirect

**Test results:**
- PASSED: 1596
- FAILED: 2

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/5/testReport/org.jbpm.bpmn2/FlowTest/testExclusiveSplitWithNoConditions/">org.jbpm.bpmn2.FlowTest.testExclusiveSplitWithNoConditions</a></summary>
Should fail as XOR gateway does not have conditions defined
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/5/testReport/org.jbpm.bpmn2.canonical/ActivityGenerationModelTest/testInclusiveSplitDefaultConnection/">org.jbpm.bpmn2.canonical.ActivityGenerationModelTest.testInclusiveSplitDefaultConnection</a></summary>
<br/>expected: 2<br/> but was: 5
</details>

**Comment by kie-ci3** [comment]
**PR job** `#8` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3459 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/8/display/redirect

**Test results:**
- PASSED: 2962
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3459/8/testReport/org.kie.kogito.serverless.workflow.executor/StaticWorkflowApplicationTest/switchFile/">org.kie.kogito.serverless.workflow.executor.StaticWorkflowApplicationTest.switchFile</a></summary>
<br/>expected: "branch1"<br/> but was: "default"
</details>

**Review by nmirasch — APPROVED** [review]

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/core/node/Split.java [hunk 1]
```diff
@@ -86,23 +85,24 @@ public boolean isDefault(final Connection connection) {
...
-            }
+            return connectionId.equals(defaultConnection);
         }
         throw new UnsupportedOperationException("Constraints are " +
                 "only supported with XOR or OR split types, not with: " + getType());
     }
 
     @Override
-    public Constraint getConstraint(final Connection connection) {
+    public Collection<Constraint> getConstraint(final Connection connection) {
```
Comment (gmunozfe) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/core/node/Split.java:105]: Same here

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/core/impl/NodeImpl.java [hunk 2]
```diff
@@ -398,7 +401,7 @@ public void setMetaData(Map<String, Object> metaData) {
         this.metaData = metaData;
     }
 
-    public Constraint getConstraint(final Connection connection) {
+    public Collection<Constraint> getConstraint(final Connection connection) {
```
Comment (gmunozfe) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/core/impl/NodeImpl.java:404]: Probably `getConstraints` in plural is more appropriated here

File: kogito-serverless-workflow/kogito-serverless-workflow-executor-core/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticWorkflowApplicationTest.java [hunk 3]
```diff
@@ -76,6 +78,19 @@ void helloWorldFile() throws IOException {
...
+    @Test
+    void switchFile() throws IOException {
+        try (
+                StaticWorkflowApplication application = StaticWorkflowApplication.create();
+                Reader reader = new InputStreamReader(Thread.currentThread().getContextClassLoader().getResourceAsStream("switch.yaml"))) {
+            Workflow workflow = ServerlessWorkflowUtils.getWorkflow(reader, WorkflowFormat.YAML);
+            assertThat(application.execute(workflow, Map.of("condition1", false, "condition2", false)).getWorkflowdata().get("output")).isEqualTo(new TextNode("default"));
+            assertThat(application.execute(workflow, Map.of("condition1", true, "condition2", false)).getWorkflowdata().get("output")).isEqualTo(new TextNode("branch1"));
+            assertThat(application.execute(workflow, Map.of("condition1", false, "condition2", true)).getWorkflowdata().get("output")).isEqualTo(new TextNode("branch1"));
+            assertThat(application.execute(workflow, Map.of("condition1", true, "condition2", true)).getWorkflowdata().get("output")).isEqualTo(new TextNode("branch1"));
```
Comment (gmunozfe) [inline kogito-serverless-workflow/kogito-serverless-workflow-executor-core/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticWorkflowApplicationTest.java:90]: check indentation

**Review by gmunozfe — APPROVED** [review]
Looks good to me, awesome work @fjtirado !
Just a minor comment about naming: `getConstraints` (in plural) returning a collection seems more accurate, but up to you.

File: kogito-serverless-workflow/kogito-serverless-workflow-executor-core/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticWorkflowApplicationTest.java (hunk 3 above)
Comment (fjtirado) [reply kogito-serverless-workflow/kogito-serverless-workflow-executor-core/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticWorkflowApplicationTest.java:90]: Thats the one the formatter generates. I agree is weird.

**Review by ricardozanini — APPROVED** [review]

**Comment by elguardian** [comment]
@fjtirado this is change does not make any sense. you could achieve this just adding new connection internally once you were creating the workflow representation. this was unrequired. and just swf was needed to change. 
Any sort of this behaviour should be query as it is jBPM and not sonata and it is shared like we have done with the internal identifiers.
Not good procedure again.
@porcelli @martinweiler

**Comment by fjtirado** [comment]
@elguardian Can you elaborate on how to achieve the solution you are proposing?
This was not proposed globally because it does not affect BPMN, so there is not procedure violation (again)

**Comment by elguardian** [comment]
@fjtirado creating two different connections with two different contraints. workflow support this sort of stuff and we need to keep to the minimum.
Again the procedure is wrong. features need to be discussed among workflow members.  You cannot make the calls on your own.

**Comment by fjtirado** [comment]
@elguardian  The engine does not support two different connection between the same two nodes. Between changing that and the number of constraints over the same connection, the call was clear to me. Anyway, point taken, I will discuss any further change in the engine (meaning jbpm-flow) no matter the impact or with you

**Pull request merged**